dask/async.py

```diff
@@ -422,8 +422,10 @@ def get_async(apply_async, num_workers, dsk, result, cache=None,
 
     if isinstance(result, list):
         result_flat = set(flatten(result))
-    else:
+    elif ishashable(result) and result in dsk:
         result_flat = set([result])
+    else:
+        return result
```

Comment:
The simplest solution here would be to assert that an invalid key will always raise `KeyError` rather than passing through.  The behavior is different right now in `core.get` and `async.get_async`.  

There is a test for `cull` that currently enforces `KeyError`:

```
assert raises(KeyError, lambda: cull(d, 'badkey'))
```
An alternative would be to guarantee that get functions work with bad keys, and create a test for that, for example:

```
def test_pass_through(self):
    d = {':x': 1,
           ':y': (inc, ':x'),
           ':z': (add, ':x', ':y')}
    assert self.get(d, 'pass-through') == 'pass-through'
    assert self.get(d, ['pass1', ['pass2', ':z']]) == ('pass1', ('pass2', 3))
```